### ISSUE TITLE
Validate custom resources without having to manually define modules for them

### ISSUE THREAD

**Issue by khaled** [hall/kubenix#36]
Currently, if custom resources are being used, they have to be specified manually using `kubernetes.customTypes`.  Furthermore, custom resources won't be validated at build time without manually defining a nix module in `kubernetes.customTypes.*.module`.  This is redundant since CRDs already define the schema.

It would be nice to not have to do any of the above.   Some potential avenues:
* Generate nix module types automatically during build, or provide a tool that does so (the former would probably require IFD?)
* Use a tool such as [kubeconform](https://github.com/yannh/kubeconform) to validate resources

**Comment by felixscheinost** [comment]
I am currently IFD to achieve that. I am not sure how I exactly I came up with it but it works for my usecase: https://gist.github.com/felixscheinost/9e7fc4370fe1ce2dfbe6efb0e75d4f4f

**Comment by adrian-gierakowski** [comment]
@felixscheinost thanks for sharing! Given that kubenix uses IFD for helm charts already I don’t think using it the way you have is a problem. I guess this could be adapted to not use IFD if the CRD schemas are known in advance.

**Comment by poelzi** [comment]
Alternatively we could use the [python-openapi](https://github.com/openapi-generators/openapi-python-client) and add some nix templates to generate nix options for any openapi spec.
The problem with kubeconform will be, to map the output to the nix input line since it will operate on the processed output. Having a nix type based definition will yield better error messages without loops and hurdles.

I need CRDs for rook, kubevirt and some other CDRs. Maybe the istio generator can be generalized enough for easier adding more CDRs.

**Comment by mibmo** [comment]
Any update on this? :)

**Comment by felixscheinost** [comment]
You are right @DerRockWolf this never really provided any type-safety. I only noticed this after some time when modules that weren't supposed to build still built. I never verified the generated .nix modules. 

I would be very interested in seeing nixidy's `crd2jsonschema.py` approach incorporated.
